connector/src/main/scala/com/microsoft/kusto/spark/utils/ContainersCache.scala

```diff
@@ -0,0 +1,42 @@
+package com.microsoft.kusto.spark.utils
+
+import com.microsoft.azure.kusto.data.Client
+import com.microsoft.kusto.spark.utils.CslCommandsGenerator.generateCreateTmpStorageCommand
+import com.microsoft.kusto.spark.utils.{KustoDataSourceUtils => KDSU}
+
+import org.joda.time.{DateTime, DateTimeZone, Period}
+
+import scala.collection.JavaConverters._
+
+class ContainersCache[A](val dmClient: Client, val clusterAlias: String, val command: String, cacheEntryCreator: ContainerAndSas => A) {
+  private var roundRobinIdx = 0
+  private var storageUris: Seq[A] = Seq.empty
+  private var lastRefresh: DateTime = new DateTime(DateTimeZone.UTC)
+  private val myName = this.getClass.getSimpleName
+
+  def getContainer = {
+    // Refresh if storageExpiryMinutes have passed since last refresh for this cluster as SAS should be valid for at least 120 minutes
+    if (storageUris.isEmpty ||
+      new Period(new DateTime(DateTimeZone.UTC), lastRefresh).getMinutes > KustoConstants.storageExpiryMinutes) {
+
+      val res = dmClient.execute(command)
+      val storage = res.getValues.asScala.map(row => {
+        val parts = row.get(0).split('?')
+        cacheEntryCreator(ContainerAndSas(parts(0), '?' + parts(1)))
```

Comment:
`cacheEntryCreator` -> `factoryMethod`?

Reply:
plus, why is this a paramater? if it is a ContainerCache, shouldn't it know how to create an entry?

Reply:
its generic .. instead i could do like the dm - create a class that would extend the cache and create the entry

Reply:
if it's generic, then the class can't be called ContainersCache...

Reply:
it is only for containers - one time export containers one time ingestion containers